api/envoy/api/v2/cds.proto

```diff
@@ -632,6 +632,25 @@ message Cluster {
   // If this flag is not set to true, Envoy will wait until the hosts fail active health
   // checking before removing it from the cluster.
   bool drain_connections_on_host_removal = 32;
+
+  // New mechanism for LB policy configuration. If set and understood
+  // by the client, replaces the lb_policy field, the common_lb_config
+  // field, and the fields in the lb_config oneof. For backward
+  // compatibility, management servers should populate both sets of fields.
+  //
+  // Every LB policy defined via this mechanism will be identified via a
+  // unique name. If the policy that needs configuration parameters, it
+  // musts define a message for its own configuration, which will be stored
+  // in the config field. The name of the policy will tell clients
+  // which type of message they should expect to see in the config field.
+  message IntraLocalityLoadBalancingPolicy {
```

Comment:
I'd like to also get a review from @vishalpowar on this.

Reply:
Thinking about this further, I don't think it's actually necessary to split up the locality-picking and endpoint-picking policies into separate fields.  Instead, we can just do this in a hierarchical fashion using nesting: if a locality-picking policy wants to allow independent selection of the endpoint-picking policy, it can do so by nesting another `LoadBalancingPolicy` message inside of its configuration proto message.

This approach is actually more similar to what we do in gRPC anyway.  I was originally thinking that the locality and endpoint picking should be specified explicitly, since the gRPC model allows routing hierarchies that are arbitrarily deep, whereas the xDS model has a hard-coded structure of endpoints within localities.  But I think even with that fixed hierarchy, it seems more flexible to allow the independent configuration of endpoint picking to be done at the discretion of the locality picking policy.

Please let me know what you think.

Reply:
I don't think we should have arbitrary deep hierarchy for LB, it should be two levels.
From the client's perspective, it will always need both Locality and Endpoint LB policy. So shouldn't we make sure the API has placeholder for both of them rather than depend on any ?

Reply:
> From the client's perspective, it will always need both Locality and Endpoint LB policy.

That's not actually true.  Most of Envoy's current built-in policies handle both locality picking and endpoint picking in a single policy.  They do not provide the ability to select the policies for locality and endpoint picking independently of each other.

Because we have use-cases where we do need to independently select the locality picking and endpoint picking policies, we know that we will need to add top-level policies that do support this in the future.  However, I would not expect Envoy's current policies to go away when we do this.  What this means is that we will be in a world where some but not all LB policies will allow independently configuring the endpoint-picking policy.

To support that, I think it makes sense to have this field be an attribute of the per-policy configuration for the parent policy.  That way, parent policies that support this will provide a field for it, and those that don't support it will not provide a field for it.  This seems cleaner than having a top-level field for specifying the endpoint-picking policy, because then it becomes possible to have an invalid config (the config specifies an endpoint-picking policy but the top-level policy does not support that feature).

> So shouldn't we make sure the API has placeholder for both of them rather than depend on any ?

This approach does not actually add any new `Any` fields.  The only `Any` field here will be the one for the per-policy LB configuration.  For top-level policies that support independently configuring endpoint-picking policy, their per-policy configuration message will directly contain another `LoadBalancingPolicy` message (i.e., it will be a direct member of the proto message, not via an `Any`).

Reply:
We definitely need to handle the current algorithms configuration that specify both locality + endpoint picking. Isin't that already covered by having the enum specifying LOAD_BALANCING_POLICY_CONFIG.

I was suggesting that we add endpoint policy explicitly specified in LoadBalancingPolicy.Policy. This will prevent misconfiguration (endpoint policy provided but no locality policy).

message LoadBalancingPolicy {
  message Policy {
 . . .
    // Optional endpoint policy, defaults to Random endpoint picks
    message EndpointPolicy {
       string name;
    } 
  }

Reply:
> We definitely need to handle the current algorithms configuration that specify both locality + endpoint picking. Isin't that already covered by having the enum specifying LOAD_BALANCING_POLICY_CONFIG.

For xDS v3, I think the goal is that we will remove the existing enum field, and all LB policies (both the existing builtin Envoy policies and any new ones we will add) will be configured using this new mechanism.

I absolutely agree that we need to support the ability to independently configure locality picking and endpoint picking.  However, not all policies will support that feature, and this configuration mechanism needs to handle all policies, regardless of whether or not they support it.

> I was suggesting that we add endpoint policy explicitly specified in LoadBalancingPolicy.Policy. This will prevent misconfiguration (endpoint policy provided but no locality policy).

I don't think the interesting case here is configuring an endpoint policy without a locality policy.  The approach in this PR would also not allow that, so the proposal you provided isn't an improvement over this PR in terms of that kind of misconfiguration.

I think the interesting case here is accidentally configuring an endpoint-picking policy when the top-level policy does not support independently configuring an endpoint-picking policy.  The approach in this PR would not allow that, because the per-policy config message for the top-level policy would simply not provide a way to set the endpoint-picking policy.  However, the proposal you provided would allow that misconfiguration, which seems worse.

> message LoadBalancingPolicy {
> message Policy {
> . . .
> // Optional endpoint policy, defaults to Random endpoint picks
> message EndpointPolicy {
> string name;
> }
> }

Note that the endpoint-picking policy will have its own config message, just as the top-level policy does.  So even if we went with this proposal, we would still need an `Any` field inside of the `EndpointPolicy` message for the config, which means that it would have the same number of `Any` fields as in the proposal in this PR.

Reply:
Thanks for the explanation.